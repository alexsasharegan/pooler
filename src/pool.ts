import { EventEmitter } from "events";
import { WrapErr } from "./wrap";

/**
 * FactoryFunc returns a Promise
 * that resolves with type T.
 */
export type FactoryFunc<T> = () => Promise<T>;
/**
 * DestructorFunc takes a type T
 * and returns a promise when it has been destroyed.
 */
export type DestructorFunc<T> = (t: T) => Promise<void>;
/**
 * HealthCheckerSync takes a value T
 * and synchronously decides if it is
 * healthy to return to the pool.
 */
export type HealthCheckerSync<T> = (t: T) => boolean;
/**
 * HealthChecker takes a value T
 * and asynchronously decides if it is
 * healthy to return to the pool.
 */
export type HealthChecker<T> = (t: T) => Promise<boolean>;

type DeferredPromise<T> = (value: T | PromiseLike<T>) => void;

/**
 * PoolOptions specify the pool's behavior.
 */
export type PoolOptions<T> = {
  /**
   * factory returns a promise that resolves with the value T.
   */
  factory: FactoryFunc<T>;
  /**
   * destructor returns a promise that resolves
   * once the value T has been destroyed.
   */
  destructor: DestructorFunc<T>;
  /**
   * is_ok returns a promise that resolves with whether or not the value T
   * is eligible to return to the pool.
   */
  is_ok?: HealthChecker<T>;
  /**
   * is_ok_sync synchronously returns whether or not the value T
   * is eligible to return to the pool.
   */
  is_ok_sync?: HealthCheckerSync<T>;
  /**
   * max is the maximum number of T you want pooled.
   */
  max?: number;
  /**
   * min is the threshold at which the pool is re-buffered.
   */
  min?: number;
  /**
   * max_retries is the number of times the factory func
   * should be retried before failing.
   */
  max_retries?: number;
  /**
   * buffer_on_start indicates whether or not the pool
   * should start buffering T upon instantiation.
   */
  buffer_on_start?: boolean;
};

/**
 * Pooler pools values of type `<T>` for reuse.
 * Values that are expensive to create (e.g. db connections)
 * can be created ahead of time and used by the application.
 */
export interface Pooler<T> {
  /**
   * Get returns a promise that resolve with a value of type `<T>`.
   * If a value is available, the promise is resolved immediately.
   * If a value is not available, the caller will be enqueued
   * and resolved in FIFO order.
   */
  Get(): Promise<T>;
  /**
   * Put returns a value of type `<T>` back to the pool.
   * If the pool is already full, the destructor for `T` will invoked
   * and the value will be discarded.
   */
  Put(x: T): Promise<void>;
  /**
   * Size returns the current size of the buffer pool;
   * the number of type `<T>` available in the pool.
   */
  Size(): number;
  /**
   * Use takes an async callback to run with a value of type `<T>`.
   * Use abstracts both Pooler.Get and Pooler.Put logic,
   * invoking the callback with `T` and
   * returning it to the pool once resolved
   * or destroying it on error.
   */
  Use(
    callback: (x: T) => Promise<void>,
    onError?: (err: any) => void
  ): Promise<void>;
  /**
   * Buffer will asynchronously buffer type `<T>` up to max.
   * The returned promise resolves once the buffering is complete.
   * If the pool is currently buffering, Buffer returns early.
   */
  Buffer(): Promise<void>;
  /**
   * Drain empties the pool and invokes the destructor on each value `<T>`.
   * It resolves once all destructors have resolved.
   */
  Drain(): Promise<void>;
}

export function NewPooler<T>(options: PoolOptions<T>): Pool<T> {
  return new Pool<T>(options);
}

export class Pool<T> extends EventEmitter implements Pooler<T> {
  private buf: T[] = [];
  private deferred: DeferredPromise<T>[] = [];
  private filling: boolean = false;
  private draining: boolean = false;
  public factory: FactoryFunc<T>;
  public destructor: DestructorFunc<T>;
  public is_ok?: HealthChecker<T>;
  public is_ok_sync?: HealthCheckerSync<T>;
  public max: number;
  public min: number;
  public max_retries: number;

  constructor(options: PoolOptions<T>) {
    // EventEmitter constructor.
    super();
    // Extract options with defaults.
    let {
      factory,
      destructor,
      is_ok,
      is_ok_sync,
      max = 10,
      min = 3,
      max_retries = 3,
      buffer_on_start = true,
    } = options;

    // Assignment.
    this.factory = factory;
    this.destructor = destructor;
    this.is_ok = is_ok;
    this.is_ok_sync = is_ok_sync;
    this.max = max;
    this.min = min;
    this.max_retries = max_retries;

    // monitorLevels will refill the pool if our minimum is reached.
    this.on("release", this.monitor_levels.bind(this));
    // flushDeferred will resolve callers awaiting a value T.
    this.on("added", this.flush_deferred.bind(this));

    if (buffer_on_start) {
      // Buffer the pool.
      this.Buffer();
    }
  }

  private async retry_factory(retries: number): Promise<T> {
    if (!retries) {
      let err = new RangeError(
        `Max attempts to create new pooled type exceeded.`
      );
      err.name = "RetryLimitError";
      throw err;
    }

    let factory_result = await WrapErr(this.factory());
    if (!factory_result.ok) {
      console.error(factory_result.error);
      return this.retry_factory(--retries);
    }

    return factory_result.value;
  }

  public async Buffer(): Promise<void> {
    if (this.filling) {
      await new Promise(r => this.once("full", r));
      return;
    }

    let fill_to = this.max - this.Size();
    if (fill_to < 1) {
      return;
    }

    const fill_one = async () => {
      let fill_result = await WrapErr(this.retry_factory(this.max_retries));
      if (!fill_result.ok) {
        return console.error(fill_result.error);
      }
      this.Put(fill_result.value);
    };

    let ps: Promise<void>[] = [];
    this.filling = true;
    this.emit("buffering");

    while (fill_to--) {
      ps.push(fill_one());
    }

    await Promise.all(ps);
    this.filling = false;
    this.emit("full");
  }

  private monitor_levels() {
    if (this.Size() < this.min && !this.draining) {
      this.Buffer();
    }
  }

  private async flush_deferred() {
    let d: DeferredPromise<T> | undefined;
    let x: T | undefined;

    // Run while we have callers waiting and buffered <T> to release.
    while (
      this.Size() > 0 &&
      (d = this.deferred.shift()) &&
      (x = this.buf.shift())
    ) {
      this.emit("release");
      d(x);
    }
  }

  public Size(): number {
    return this.buf.length;
  }

  public async Get(): Promise<T> {
    // Release a value <T> from the pool.
    // Shift off the front since we'll push on to the end.
    // Should keep values from getting too stale.
    let x = this.buf.shift();
    if (!x) {
      // Create a promise to be resolved once the buffer refills.
      // Internal 'added' event listener will flush these deferred promises.
      return new Promise<T>(resolve => {
        this.deferred.push(resolve);
      });
    }

    // Now that we're guaranteed a value, emit our release event.
    this.emit("release");
    return x;
  }

  public async Put(x: T): Promise<void> {
    if (this.Size() >= this.max || this.draining) {
      return this.destructor(x);
    }

    if (this.is_ok_sync && !this.is_ok_sync(x)) {
      return this.destructor(x);
    }

    if (this.is_ok && !await this.is_ok(x)) {
      return this.destructor(x);
    }

    let y: T;
    for (y of this.buf) {
      if (x === y) {
        throw new TypeError(`Cannot 'Put' duplicate value.`);
      }
    }

    // Added values are pushed on the end
    // since we retrieve from the start.
    this.buf.push(x);
    this.emit("added");
  }

  public async Use(
    callback: (x: T) => Promise<void>,
    onError?: (err: any) => void
  ): Promise<void> {
    let x = await this.Get();
    let cb_result = await WrapErr(callback(x));

    if (!cb_result.ok) {
      console.error(cb_result.error);
      this.destructor(x);
      if (onError) {
        onError(cb_result.error);
      }
      return;
    }

    this.Put(x);
  }

  public async Drain(): Promise<void> {
    if (this.draining) {
      await new Promise(r => this.once("drained", r));
      return;
    }

    let ps: Promise<void>[] = [];
    let x: T | undefined;

    this.draining = true;

    while ((x = this.buf.shift())) {
      ps.push(this.destructor(x));
    }

    await WrapErr(Promise.all(ps));

    this.draining = false;
    this.emit("drained");
  }
}
