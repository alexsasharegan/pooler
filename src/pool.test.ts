import { Pooler, NewPooler } from "./pool";
import { setTimeout } from "timers";
import { promisify } from "util";

const wait = promisify(setTimeout);
const mock_delay = 10;
const mock_name = "PoolMock Object";
const mock_opts = () => ({
  factory: PoolMock.factory,
  destructor: PoolMock.destructor,
  max: 10,
  min: 2,
});

class PoolMock {
  public name: string = mock_name;

  public static async factory(): Promise<PoolMock> {
    await wait(mock_delay);
    return new PoolMock();
  }

  public static async destructor(_mock: PoolMock): Promise<void> {
    await wait(mock_delay);
  }
}

describe("Pool Public API:", () => {
  it("should get a value `<T>` from the pool", async () => {
    let pool: Pooler<PoolMock>;
    let mock: PoolMock;

    pool = await NewPooler<PoolMock>(mock_opts());
    mock = await pool.get();
    expect(mock).toBeDefined();
    expect(mock.name).toBe(mock_name);
  });

  it("should put a value `<T>` into the pool", async () => {
    let opts = Object.assign(mock_opts(), { buffer_on_start: false });
    let pool = await NewPooler<PoolMock>(opts);
    let mock = new PoolMock();

    expect(pool.size()).toBe(0);
    await pool.put(mock);
    expect(pool.size()).toBe(1);
  });

  it("should not put the same value `<T>` into the pool twice", async () => {
    let opts = Object.assign(mock_opts(), { buffer_on_start: false });
    let pool = await NewPooler<PoolMock>(opts);
    let mock = new PoolMock();

    expect(pool.size()).toBe(0);
    await pool.put(mock);
    expect(pool.put(mock)).rejects.toThrow(TypeError);
  });

  it("should not put a value `<T>` into the pool when at max", async () => {
    let spy = jest.fn();
    let opts = Object.assign(mock_opts(), { destructor: spy });
    let pool = await NewPooler<PoolMock>(opts);

    expect(pool.size()).toBe(opts.max);
    expect(spy).not.toHaveBeenCalled();

    let pm = new PoolMock();
    await pool.put(pm);
    expect(pool.size()).toBe(opts.max);
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should not buffer when at max", async () => {
    let opts = Object.assign(mock_opts());
    let pool = await NewPooler<PoolMock>(opts);

    expect(pool.size()).toBe(opts.max);

    await pool.buffer();
    expect(pool.size()).toBe(opts.max);
  });

  it("should drain", async () => {
    let spy = jest.fn();
    let opts = Object.assign(mock_opts(), { destructor: spy });
    let pool = await NewPooler<PoolMock>(opts);

    expect(spy).not.toHaveBeenCalled();

    await pool.drain();
    expect(pool.size()).toBe(0);
    expect(spy).toHaveBeenCalledWith(expect.any(PoolMock));
    expect(spy).toHaveBeenCalledTimes(opts.max);
  });

  it("should handle calling Drain twice", async () => {
    let spy = jest.fn();
    let opts = Object.assign(mock_opts(), { destructor: spy });
    let pool = await NewPooler<PoolMock>(opts);
    let ps: Promise<any>[] = [];

    expect(spy).not.toHaveBeenCalled();

    ps.push(pool.drain().then(() => expect(pool.size()).toBe(0)));
    ps.push(pool.drain().then(() => expect(pool.size()).toBe(0)));
    await Promise.all(ps);
    expect(spy).toHaveBeenCalledWith(expect.any(PoolMock));
    expect(spy).toHaveBeenCalledTimes(opts.max);
  });
});

describe("Pool options", () => {
  it("should call 'factory'", async () => {
    // Use a Jest spy, but wrap it so we can return unique objects for the Pool.
    let spy = jest.fn();
    let mock_spy = () => (spy(), new PoolMock());
    let opts = Object.assign(mock_opts(), { factory: mock_spy });
    let pool = await NewPooler<PoolMock>(opts);

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenCalledTimes(opts.max);
  });

  it("should call 'destructor'", async () => {
    let spy = jest.fn();
    let opts = Object.assign(mock_opts(), { destructor: spy });
    let pool = await NewPooler<PoolMock>(opts);

    expect(spy).not.toHaveBeenCalled();

    await pool.drain();
    expect(spy).toHaveBeenCalledTimes(opts.max);
  });

  it("should use 'buffer_on_start'", async () => {
    let opts = Object.assign(mock_opts(), { buffer_on_start: false });
    let pool = await NewPooler<PoolMock>(opts);

    // Since pool is buffered asynchronously,
    // wait to allow our pool a chance to buffer.
    await wait(mock_delay * 2);
    expect(pool.size()).toBe(0);

    await pool.buffer();
    expect(pool.size()).toBe(opts.max);

    opts.buffer_on_start = true;
    pool = await NewPooler<PoolMock>(opts);

    expect(pool.size()).toBe(opts.max);
  });

  it("should use 'max'", async () => {
    let max = 25;
    let opts = Object.assign(mock_opts(), { max });
    let pool = await NewPooler<PoolMock>(opts);

    // Pool should begin buffering immediately,
    // but not fill until at least `MockDelay` has elapsed.
    expect(pool.size()).toBe(max);

    // Shouldn't overfill
    await pool.buffer();
    expect(pool.size()).toBe(max);

    max = 5;
    opts = Object.assign(mock_opts(), { max });
    pool = await NewPooler<PoolMock>(opts);

    expect(pool.size()).toBe(max);
  });
});

describe("Pool internals", () => {
  it("should re-buffer when the 'min' threshold is met", async () => {
    let factory_spy = jest.fn();
    let opts = Object.assign(mock_opts(), {
      factory() {
        factory_spy();
        return PoolMock.factory();
      },
    });
    let count = 0;
    let pool = await NewPooler<PoolMock>(opts);

    expect(factory_spy).toHaveBeenCalledTimes(opts.max);
    count += opts.max;

    // Cross the threshold by just one.
    let i = opts.max - opts.min + 1;
    count += i;
    while (i--) {
      expect(await pool.get()).toBeInstanceOf(PoolMock);
    }

    await wait(mock_delay * 5);

    expect(factory_spy).toHaveBeenCalledTimes(count);
    expect(pool.size()).toBe(opts.max);
  });

  it("should flushed deferred callers", async () => {
    let opts = mock_opts();
    let pool = await NewPooler<PoolMock>(opts);

    // Brute force pool drain.
    for (let i = 0; i < opts.max; i++) {
      pool.get().then(x => expect(x).toBeInstanceOf(PoolMock));
    }

    expect(pool.size()).toBe(0);
    expect(await pool.get()).toBeInstanceOf(PoolMock);
  });

  it("should retry the factory function", async () => {
    let count = 0;
    let throw_limit = 1;
    let max_retries = 5;
    let message = "Retry test.";
    let opts = Object.assign(mock_opts(), {
      max_retries,
      async factory() {
        if (count < throw_limit) {
          count++;
          throw new Error(message);
        }

        return new PoolMock();
      },
    });
    let pool = await NewPooler<PoolMock>(opts);

    expect(count).toBe(throw_limit);
  });

  it("should fail when max_retries exceeded in factory", async () => {
    let message = "Retry test.\n\nThis is OK.\n";
    let opts = Object.assign(mock_opts(), {
      buffer_on_start: false,
      max_retries: 1,
      async factory() {
        throw new Error(message);
      },
    });

    try {
      let pool = await NewPooler<PoolMock>(opts);
      await pool.buffer();
    } catch (err) {
      expect(err).toBeInstanceOf(RangeError);
    }
  });
});
