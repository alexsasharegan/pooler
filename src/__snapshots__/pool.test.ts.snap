// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pool internals should use default options 1`] = `10`;
